### PR TITLE
Ble client additions and fixes

### LINKED ISSUE #4802: BLE Sensor notify config flag not working
### The problem

The BLE Client Sensor component has a config option `notify` which should turn on notifications for the specified characteristic. This is not working (and thus the `on_notify` automation is never triggered) due to two typos in `ble_client_base.cpp`.

I have a fix for this and will include it in an upcoming PR.

### Which version of ESPHome has the issue?

2023.9.0-dev

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

N/A

### What platform are you using?

ESP32

### Board

nodemcu-32s

### Component causing the issue

BLE Client Sensor

### Example YAML snippet

```yaml
sensor:
  - platform: ble_client
    type: characteristic
    ble_client_id: door
    name: Temperature
    service_uuid: '181A'
    characteristic_uuid: '2A6E'
    unit_of_measurement: '<C2><BA>C'
    accuracy_decimals: 1
    update_interval: 10s
    notify: true
    on_notify:
      then:
        - logger.log: "Notify event"
    lambda: |-
      return (float)(x[0] + (x[1] << 8)) / 100.0f;
```

### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

- Add new actions for the ble_client component - `ble_client.connect` and `ble_client.disconnect`;
- Add new config option for ble_client - `auto_connect`;
- Fix (at least) one bug;
- Use `play_complex()` and `play_next()` in the new actions, and the existing `ble_client.ble_write` action so that they can be used sequentially in automation blocks.
- General cleanup and refactor of the `ble_client` code with particular attention to correct handling of events, especially when multiple actions/triggers are active on the same client.

The `ble_client` modules for `sensor`, `binary_sensor`, `switch` and the various triggers have been tested on hardware. Other components for specific hardware that rely on the `ble_client` component have not been tested by me due to lack of access to that hardware, but compilation has been tested for a selection of these, and a careful code review has been performed to check for potential breaking changes in the `BLEClient` class.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes [https://github.com/esphome/issues/issues/4802](https://github.com/esphome/issues/issues/4802)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3138

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
ble_client:
  - id: ble_clock
    mac_address: 17:75:BC:F2:94:4D
    auto_connect: false
  - id: door
    mac_address: 00:0D:6F:F6:1C:B3
  - id: blout
    mac_address: 68:b6:b3:21:69:92

time:
  - platform: sntp
    id: sntp_time
    timezone: Australia/Sydney

interval :
  - interval: 60min
    then:
      - if:
          condition:
            time.has_time:
          then:
            - esp32_ble_tracker.stop_scan:
            - ble_client.connect: ble_clock
            - ble_client.ble_write:
                id: ble_clock
                service_uuid: EBE0CCB0-7A0A-4B0C-8A1A-6FF2997DA3A6
                characteristic_uuid: EBE0CCB7-7A0A-4B0C-8A1A-6FF2997DA3A6
                value: !lambda |-
                    uint32_t t = id(sntp_time).now().timestamp + ESPTime::timezone_offset();
                    return {(uint8_t)t, (uint8_t)(t >> 8), (uint8_t)(t >> 16), (uint8_t)(t >> 24), 0};
            - ble_client.disconnect: ble_clock
            - esp32_ble_tracker.start_scan:
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by clydebarrow** [comment]
Not stale, just waiting for review.

**Comment by clydebarrow** [comment]
Still not stale - just old :-)

**Comment by 2m** [comment]
Eagerly waiting for this. :) Currently it is not convenient to control Valve Basestations with ESPHome, because when ESPHome is connected, then my PC can not see base stations at all.

File: esphome/components/ble_client/automation.h [hunk 1]
```diff
@@ -126,10 +116,94 @@ template<typename... Ts> class BLEClientWriteAction : public Action<Ts...>, publ
...
+  }
+
+  /**
+   * Note about logging: the esph_log_X macros are used here because the CI checks complain about use of the ESP LOG
+   * macros in header files (Can't even write it in a comment!)
+   * Not sure why, because they seem to work just fine.
+   * The problem is that the implementation of a templated class can't be placed in a .cpp file when using C++ less than
+   * 17, so the methods have to be here.  The esph_log_X macros are equivalent in function, but don't trigger the CI
+   * errors.
+   */
```
Comment (jesserockz) [inline esphome/components/ble_client/automation.h:137]: The check was added in https://github.com/esphome/esphome/pull/546
But there is no explanation as to why

**Comment by tvvalbonne** [comment]
Any chance to get this reviewed? waiting for it...

**Review by kbx81 — APPROVED** [review]

**Pull request merged**